workers/unity/Packages/io.improbable.worker.sdk/EventTracing.cs

```diff
@@ -0,0 +1,349 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using Improbable.Worker.CInterop.Internal;
+
+namespace Improbable.Worker.CInterop
+{
+    public interface ITraceItem
+    {
+    }
+
+    public unsafe struct SpanId
+    {
+        public fixed byte Data[16];
+    }
+
+    public struct Span : ITraceItem
+    {
+        public SpanId Id;
+        public int CauseCount;
+        public SpanId[] Causes;
+    }
+
+    public struct Event : ITraceItem
+    {
+        public SpanId Id;
+        public ulong UnixTimestampMillis;
+        public string Message;
+        public string Type;
+        public TraceEventData Data;
+    }
+
+    public enum ItemType
+    {
+        Span = 1,
+        Event = 2
+    }
+
+    public struct Item
+    {
+        public ItemType ItemType;
+        public ITraceItem TraceItem;
+    }
+
+    public static class TraceSpan
+    {
+        public static byte SpanIdHash(SpanId spanId)
+        {
+            return CEventTrace.SpanIdHash(ParameterConversion.ConvertSpanId(spanId));
+        }
+
+        public static bool SpanIdEqual(SpanId spanId1, SpanId spanId2)
+        {
+            return CEventTrace.SpanIdEqual(
+                ParameterConversion.ConvertSpanId(spanId1),
+                ParameterConversion.ConvertSpanId(spanId2)) > 0;
+        }
+
+        public static unsafe SpanId GetNullSpanId()
```

Comment:
The function needed to be marked as `unsafe` since it uses the `Data` property which is a pointer to the array. Once I moved it inside the `SpanId` struct that is marked as `unsafe`, I could remove the `unsafe` modifier from the function.

If I wanted to remove `unsafe` from the SpanId struct, would I need to introduce a wrapper around the struct?

Reply:
You can also have a standalone `unsafe { }` block inside a safe function like: 

```csharp
public static SpanId GetNullSpanId()
{
	unsafe 
	{
		// Do unsafe things in here. By doing so, we are saying that we've ensured the invariants
		// of the unsafe code are met and calling this function is now 'safe'. 
	} 
}
```

Reply:
Ohh right, I forgot you could do that! Ty!

Reply:
Oh also, since the `Data` field is fixed, is it okay to do:

```csharp
public unsafe fixed byte Data[16];
```